ls1intum/Artemis
# Add Option to Disable More Feedback Requests

### Checklist
- [ ] The database changes got approved. --- **Do not test this PR on the test servers until this is checked!**

### Motivation and Context
Some courses need to be able to adjust the assessment after a complaint if it is wrong because exam-rules apply, practical courses such as PGdP have this requirement. Student using the more feedback function only make everything more complicated, 
because the score cannot be modified by tutors. We want to diable more feedback requests for the course and only allow (unlimited) complaints because we more or less have to.

### Description
- **Database Change: added `request_more_feedback_enabled` column**
  It is not nullable and a primitive boolean.
- Added a corresponding attribute to the `Course` class and adjusted some logic
- Changed course and complaint validation (on create/update)
- Added the attribute on the client side as well
- Adjust client forms and some complaint-setting-based conditions

#### TODO
- [ ] Find out why team complaints do not work at all (even on develop, or am I doing something wrong?)
  Seems to be decided by `this.isCurrentUserSubmissionAuthor = this.participation.student.id === user.id;`, is that a problem?
- [ ] The same goes for more feedback requests for teams, those don't seem to exist as well.

### Steps for Testing
0. Review the code, especially look if I missed something that needs to be changed as well
1. Test that complaints with various course settings work:
    1. Request more feedback enabled <-> disabled
    2. Positive number of max complains <-> zero
    3. Positive number of max TEAM complains <-> zero
    4. Or disabling complaints completely (more feedback requests are a special type of complaint, so they are disabled as well).
2. Adjust the course setting and then see, if the correct options are available for an _already graded submission in this course_.
    "Request More Feedback" should only be available if the option is checked, and "Complain" only when complaint tokens are available.

### Test Coverage
_TODO_

### Screenshots
- **Options when complaints enabled**
  ![image](https://user-images.githubusercontent.com/11130248/96325905-7c015680-102c-11eb-9112-bd9a512cf1a5.png)
- **Option hidden when no complaints at all**
  ![image](https://user-images.githubusercontent.com/11130248/96325910-8d4a6300-102c-11eb-8fb9-dd9624c3eaca.png)
- **Students can only submit complaints and not request more feedback if disabled**
  ![image](https://user-images.githubusercontent.com/11130248/96325931-c682d300-102c-11eb-9998-6d8b4ea22525.png)
